### PR TITLE
fix(storage): harden system settings path override (Fixes #1987)

### LINKED ISSUE #1987: Harden Storage.getSystemSettingsPath(): validate LLXPRT_SYSTEM_SETTINGS_PATH env override
Tracking follow-up from PR #1982 (storage extraction, deferred to keep behavior identical).

In packages/storage/src/config/storage.ts, getSystemSettingsPath() returns the LLXPRT_SYSTEM_SETTINGS_PATH environment variable directly. Consider validating/normalizing it (path.resolve / allowlist) to guard against path traversal.

Pre-existing behavior; intentionally not changed in the extraction PR.

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Harden the storage system settings path environment override so LLXPRT_SYSTEM_SETTINGS_PATH is normalized before use and relative or blank overrides fall back to the platform default.

## Dive Deeper

This change adds a small validation boundary in Storage.getSystemSettingsPath:

- absolute override paths are returned as resolved absolute paths
- relative override paths are ignored so traversal cannot be returned directly from the environment
- blank or whitespace-only override values are ignored
- Storage.getSystemPoliciesDir continues to derive from getSystemSettingsPath, so it now uses the sanitized settings path too

Behavioral tests cover default paths, normalized absolute overrides, traversal segment collapse, rejection of relative and blank overrides, and policy directory derivation from the sanitized path.

## Reviewer Test Plan

Reviewers can validate by setting LLXPRT_SYSTEM_SETTINGS_PATH to:

- an absolute path containing parent-directory segments and confirming Storage.getSystemSettingsPath returns the resolved path
- a relative path such as ../settings.json and confirming the platform default is returned
- a whitespace-only value and confirming the platform default is returned

Automated validation run locally:

- npm run test --workspace @vybestack/llxprt-code-storage -- src/config/storage.test.ts
- npm run lint --workspace @vybestack/llxprt-code-storage
- npm run test
- npm run lint
- npm run typecheck
- npm run format
- npm run build
- node scripts/start.js --profile-load ollamakimi "write me a haiku and nothing else"

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Fixes #1987

### CODE REVIEW

File: packages/storage/src/config/storage.ts [hunk 1]
```diff
@@ -14,6 +14,16 @@ export const PROVIDER_ACCOUNTS_FILENAME = 'provider_accounts.json';
...
 
+function resolveSystemSettingsEnv(raw: string | undefined): string | undefined {
+  if (raw === undefined || raw.trim() === '') {
+    return undefined;
+  }
+  if (!path.isAbsolute(raw)) {
+    return undefined;
+  }
+  return path.resolve(raw);
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/storage/src/config/storage.ts:29]: Addressed in fcce36a04 by trimming the environment override once and reusing the trimmed value for the empty check, absolute-path validation, and path resolution. Added behavioral coverage for leading, trailing, and surrounding whitespace around absolute overrides.

**Comment by acoliver** [comment]
## Automated Review — LLxprt Code + Open Code Review (OCR)

> This comment was generated by **LLxprt Code** (automated agent) running the **codeanalyzer** subagent, independent source verification, and the **Open Code Review (OCR)** CLI (`ocr review`, 15-min/file timeout, test files included). It is **not** authored by the PR author.

---

### Verdict: ✅ Approve — the fix is genuine, no cheating/masking detected

PR #2176 correctly resolves issue #1987. Below is the collated analysis from three independent review passes.

---

### 1. Does it actually fix #1987?

**Yes.** The original code returned `LLXPRT_SYSTEM_SETTINGS_PATH` raw and unsanitized:

    // BEFORE (main)
    static getSystemSettingsPath(): string {
      if (process.env['LLXPRT_SYSTEM_SETTINGS_PATH']) {
        return process.env['LLXPRT_SYSTEM_SETTINGS_PATH']; // raw, unnormalized
      }

The fix adds a `resolveSystemSettingsEnv()` helper that:

- **Rejects** `undefined`, blank, and whitespace-only values → falls back to platform default
- **Rejects** relative paths (`../settings.json`, `relative/settings.json`) → falls back to platform default
- **Normalizes** absolute paths via `path.resolve()`, collapsing `..` and `.` segments

    // AFTER (this PR)
    const sanitized = resolveSystemSettingsEnv(
      process.env['LLXPRT_SYSTEM_SETTINGS_PATH'],
    );
    if (sanitized !== undefined) {
      return sanitized;
    }

**No cheating detected.** A full scan of the diff found **zero** `eslint-disable`, `@ts-ignore`, `@ts-expect-error`, `@ts-nocheck`, or any other rule-suppression patterns introduced by this PR. The production code is genuinely fixed, not masked.

**Coverage is complete.** The env var `LLXPRT_SYSTEM_SETTINGS_PATH` is read in **exactly one place** (`storage.ts:84`). All downstream consumers — `settings.ts`, `settingsLoader.ts`, and `getSystemPoliciesDir()` — go through the now-sanitized `getSystemSettingsPath()`, so the fix propagates everywhere.

---

### 2. OCR Findings (4 comments, 3 files reviewed, ~357K tokens, 1m48s)

OCR found **0 high-severity** issues. All 4 findings are test-quality / maintainability improvements:

#### Finding A — Missing empty-string edge-case test (Medium)

**`storage.test.ts:248`** — There is a test for whitespace-only (`'   '`) but **no test for empty-string** (`''`). Although both hit the same `trimmed === ''` branch, an explicit empty-string test documents the contract and guards against future regressions. Suggested addition:

    it('ignores an empty-string override in favor of platform default', () => {
      process.env[ENV_KEY] = '';
      expect(Storage.getSystemSettingsPath()).toBe(
        expectedDefaultSystemSettingsPath(),
      );
    });

#### Finding B — Missing `numRuns` cap on setProvider property test (Medium)

**`mutationCoverage.behavior.test.ts:415-427`** — The first property test in this block correctly added `{ numRuns: 5 }` to stay within CI timeout budgets. However, the `setProvider` model-preservation property test was left with the **default `numRuns` (100)**. Under the fake seam, `listProviders()` returns only `['fake']`, so `fc.constantFrom('fake')` generates the same value every iteration — meaning 100 identical agent build/teardown cycles with no added coverage. This risks the same 30s timeout flakiness the PR is trying to eliminate. Suggested fix:

    await fc.assert(
      fc.asyncProperty(fc.constantFrom(...names), async (providerName) => {
        // ...
      }),
      { numRuns: Math.max(names.length, 1) },
    );

#### Finding C — Duplicated platform default paths in test helper (Medium)

**`storage.test.ts:26-34`** — `expectedDefaultSystemSettingsPath()` hardcodes the same platform-specific paths as `storage.ts`. If the defaults in `storage.ts` ever change, these test expectations will silently diverge. Consider exporting a `getDefaultSystemSettingsPath()` from `storage.ts` and importing it here.

*Note from LLxprt Code:* There is a legitimate tension here — importing the implementation's default makes the "returns platform default" assertion tautological (you'd be asserting the function returns what the function returns). The current approach (testing against known-good hardcoded values) is defensible. A follow-up could export the defaults as named constants and assert against those.

#### Finding D — Three whitespace tests could be consolidated (Low)

**`storage.test.ts:263-279`** — The leading/trailing/surrounding whitespace tests are near-duplicates. They could be consolidated into a single table-driven `it.each` test. This is purely a readability suggestion; the current explicit tests are correct and clear.

---

### 3. Codeanalyzer Subagent Assessment

The codeanalyzer independently confirmed:
- The fix is **genuine and complete** — no masking or cheating
- Test changes to `mutationCoverage.behavior.test.ts` are **legitimate corrections**, not coverage weakening:
  - `fc.string()` → `fc.constantFrom(...names)` correctly constrains inputs to registered providers (the old test was generating arbitrary strings that would always throw "Provider not found")
  - `{ numRuns: 5 }` on the handler-divergence test is justified (2 agents per run × 100 runs was too slow for CI)
- `getSystemPoliciesDir()` properly derives from the sanitized path
- No symlink resolution, but this is acceptable given the threat model (env-var control implies administrative trust — noted by CodeRabbit in the issue)

---

### 4. Remaining Considerations (informational, not blockers)

- **Absolute-path traversal**: An attacker who controls the env var can still point to any absolute path (e.g., `/etc/shadow`). `path.resolve()` collapses `..` segments but does not implement an allowlist. This is acceptable per the threat model — if an attacker can set environment variables, they already have significant system access. The issue explicitly offered "path.resolve / allowlist" as options; this PR chose the lighter-touch `path.resolve` approach, which is appropriate.
- **No symlink handling**: A symlinked settings path would resolve to its target. Not addressed, but low-risk for the same trust-boundary reason.

---

### Summary

| Check | Result |
|-------|--------|
| Fixes #1987 genuinely? | ✅ Yes |
| No lint/rule suppression cheating? | ✅ Confirmed clean |
| Production code correct? | ✅ Yes |
| Test changes legitimate? | ✅ Yes (not weakened) |
| OCR high-severity issues? | 0 |
| OCR actionable findings | 2 Medium (empty-string test, numRuns cap), 1 Medium (dup paths), 1 Low (consolidate tests) |

**Recommendation:** Merge as-is. Findings B and A are worth addressing in a quick follow-up commit (both are one-liners). Findings C and D are optional improvements.

---

*Generated by LLxprt Code automated review pipeline: codeanalyzer subagent + independent source verification + Open Code Review (OCR). Review performed on branch `issue1987` at commit `21f9c1439`.*
